include/linux/blkdev_compat.h

```diff
@@ -301,68 +301,113 @@ bio_set_flags_failfast(struct block_device *bdev, int *flags)
  * allow richer semantics to be expressed to the block layer.  It is
  * the block layers responsibility to choose the correct way to
  * implement these semantics.
- *
- * The existence of these flags implies that REQ_FLUSH an REQ_FUA are
- * defined.  Thus we can safely define VDEV_REQ_FLUSH and VDEV_REQ_FUA
- * compatibility macros.
- *
- * Linux 4.8 renamed the REQ_FLUSH to REQ_PREFLUSH but there was no
- * functional change in behavior.
  */
 #ifdef WRITE_FLUSH_FUA
-
 #define	VDEV_WRITE_FLUSH_FUA		WRITE_FLUSH_FUA
-#ifdef REQ_PREFLUSH
-#define	VDEV_REQ_FLUSH			REQ_PREFLUSH
-#define	VDEV_REQ_FUA			REQ_FUA
 #else
-#define	VDEV_REQ_FLUSH			REQ_FLUSH
-#define	VDEV_REQ_FUA			REQ_FUA
+#define	VDEV_WRITE_FLUSH_FUA		WRITE_BARRIER
 #endif
 
+/*
+ * 4.8 - 4.x API,
+ *   REQ_OP_FLUSH
+ *
+ * 4.8-rc0 - 4.8-rc1,
+ *   REQ_PREFLUSH
+ *
+ * 2.6.36 - 4.7 API,
+ *   REQ_FLUSH
+ *
+ * 2.6.x - 2.6.35 API,
+ *   HAVE_BIO_RW_BARRIER
+ */
+static inline boolean_t
+bio_is_flush(struct bio *bio)
+{
+#if defined(HAVE_REQ_OP_FLUSH)
+	return (bio_op(bio) == REQ_OP_FLUSH);
+#elif defined(REQ_PREFLUSH)
+	return (bio->bi_rw & REQ_PREFLUSH);
+#elif defined(REQ_FLUSH)
+	return (bio->bi_rw & REQ_FLUSH);
+#elif defined(HAVE_BIO_RW_BARRIER)
+	return (bio->bi_rw & (1 << BIO_RW_BARRIER));
 #else
-
-#define	VDEV_WRITE_FLUSH_FUA		WRITE_BARRIER
-#ifdef HAVE_BIO_RW_BARRIER
-#define	VDEV_REQ_FLUSH			(1 << BIO_RW_BARRIER)
-#define	VDEV_REQ_FUA			(1 << BIO_RW_BARRIER)
-#else
-#define	VDEV_REQ_FLUSH			REQ_HARDBARRIER
-#define	VDEV_REQ_FUA			REQ_FUA
+#error	"Allowing the build will cause flush requests to be ignored. Please "
+	"file an issue report at: https://github.com/zfsonlinux/zfs/issues/new"
 #endif
+}
 
+/*
+ * 4.8 - 4.x API,
+ *   REQ_FUA flag moved to bio->bi_opf
+ *
+ * 2.6.x - 4.7 API,
+ *   REQ_FUA
+ */
+static inline boolean_t
+bio_is_fua(struct bio *bio)
+{
+#if defined(HAVE_BIO_BI_OPF)
+	return (bio->bi_opf & REQ_FUA);
+#elif defined(REQ_FUA)
+	return (bio->bi_rw & REQ_FUA);
+#else
+#error	"Allowing the build will cause fua requests to be ignored. Please "
+	"file an issue report at: https://github.com/zfsonlinux/zfs/issues/new"
 #endif
+}
 
 /*
- * 2.6.28 - 2.6.35 API,
- *   BIO_RW_DISCARD
+ * 4.8 - 4.x API,
+ *   REQ_OP_DISCARD
  *
  * 2.6.36 - 4.7 API,
  *   REQ_DISCARD
  *
- * 4.8 - 4.x API,
- *   REQ_OP_DISCARD
+ * 2.6.28 - 2.6.35 API,
+ *   BIO_RW_DISCARD
  *
  * In all cases the normal I/O path is used for discards.  The only
  * difference is how the kernel tags individual I/Os as discards.
  */
-#ifdef QUEUE_FLAG_DISCARD
 static inline boolean_t
 bio_is_discard(struct bio *bio)
 {
-#if defined(HAVE_BIO_RW_DISCARD)
-	return (bio->bi_rw & (1 << BIO_RW_DISCARD));
+#if defined(HAVE_REQ_OP_DISCARD)
+	return (bio_op(bio) == REQ_OP_DISCARD);
```

Comment:
Then you should add bio_is_secure_erase to zvol_request where it calls zvol_discard.